### PR TITLE
Better toolchain downloader

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [ ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [ ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [ ] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [ ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] If you are adding/removing a .spec file that has multiple-versions supported, please add [@microsoft/cbl-mariner-multi-package-reviewers](https://github.com/orgs/microsoft/teams/cbl-mariner-multi-package-reviewers) team as reviewer [(Eg. golang has 2 versions 1.18, 1.21+)](https://github.com/microsoft/azurelinux/tree/2.0/SPECS/golang)
- [ ] Ready to merge

---

###### Summary 

In preparation for a new explicit validation mechanism break out the toolchain downloader logic into a real script.

The new script will handle the logic that was previously baked into the makefile, while also addressing the issue of log overwriting.

###### Change Log  

- Move toolchain download logic into `download_toolchain_rpms.sh`

###### Does this affect the toolchain?  

**YES**

###### Associated issues  

- #xxxx

###### Links to CVEs  
- https://nvd.nist.gov/vuln/detail/CVE-YYYY-XXXX

###### Test Methodology

- Pipeline build id: xxxx

### CODE REVIEW

File: toolkit/scripts/toolchain/download_toolchain_rpm.sh [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    echo "-t|--downloader-tool: Path to our go downloader tool"
+    echo "-r|--rpm-name: Name of the RPM to download"
+    echo "-d|--dst: Destination file path"
+    echo "-l|--log-base: Base path for log files. Each attempt will have a log file named <log_base>.<attempt_num>"
+    echo "--hydrate: Print alternate log message for hydrate operations"
+    echo "-h|--help: Print this help message"
+    echo "-u|--url-list: Space separated list of base URLs to attempt to download the RPM from. The URLs will be tried in order"
+    echo "           until the RPM is successfully downloaded. Full URL will be <url>/<rpm_name>"
+    echo "-c|--certificate: Optional path to a certificate file to use for the download"
+    echo "-k|--private-key: Optional path to a private key file to use for the download"
```
Comment (anphel31) [inline toolkit/scripts/toolchain/download_toolchain_rpm.sh:21]: does not match usage output above (--key vs --private-key, and --certificate vs --cert)

File: toolkit/scripts/toolchain.mk [hunk 2]
```diff
@@ -296,26 +292,24 @@ $(toolchain_rpms): $(TOOLCHAIN_MANIFEST) $(STATUS_FLAGS_DIR)/toolchain_local_tem
...
-		tail -n$(toolchain_log_tail_length) $$log_file | sed 's/^/\t/' && \
-		$(call print_error,\nToolchain download failed. See above errors for more details.) \
+$(toolchain_rpms): $(TOOLCHAIN_MANIFEST) $(STATUS_FLAGS_DIR)/daily_build_auto_cleanup.flag $(depend_REBUILD_TOOLCHAIN) $(go-downloader) $(SCRIPTS_DIR)/toolchain/download_toolchain_rpm.sh
+	@log_file="$(toolchain_downloads_logs_dir)/$(notdir $@).log" && \
+	rm -f "$$log_file" && \
+	$(SCRIPTS_DIR)/toolchain/download_toolchain_rpm.sh \
+		--downloader-tool "$(go-downloader)" \
+		--rpm-name "$(notdir $@)" \
+		--dst "$@" \
+		--log-base "$$log_file" \
```
Comment (anphel31) [inline toolkit/scripts/toolchain.mk:302]: is this intentional? (`$$log_file`)

**Review by anphel31 — APPROVED** [review]

File: toolkit/scripts/toolchain.mk (hunk 2 above)
Comment (dmcilvaney) [reply toolkit/scripts/toolchain.mk:302]: Yes, its referencing the bash variable `$log_file` but needs an escape for Make.

File: toolkit/scripts/toolchain/download_toolchain_rpm.sh (hunk 1 above)
Comment (dmcilvaney) [reply toolkit/scripts/toolchain/download_toolchain_rpm.sh:21]: Fixed

**Pull request merged**